File: dclib/core/src/main/java/dclib/util/Timer.java
package dclib.util;

public final class Timer {

	private float maxTime;
	private float time;

	public Timer(final float maxTime) {
		this(maxTime, 0);
	}

	public Timer(final float maxTime, final float time) {
		this.maxTime = maxTime;
		this.time = time;
	}

	public final float getMaxTime() {
		return maxTime;
	}

	public final void setMaxTime(final float maxTime) {
		this.maxTime = maxTime;
	}

	public final float getRemainingTime() {
		return maxTime - time;
	}

	public final boolean isElapsed() {
		return time >= maxTime;
	}

	public final float getElapsedPercent() {
		return time / maxTime;
	}

	public final void reset() {
		time = 0;
	}

	public final boolean check() {
		boolean isElapsed = isElapsed();
		if (isElapsed) {
			reset();
		}
		return isElapsed;
	}

	public final void tick(final float delta) {
		time = Math.min(time + delta, maxTime);
	}

}

File: dclib/core/src/main/java/dclib/util/Timer.kt
package dclib.util

class Timer @JvmOverloads constructor(var maxTime: Float = 0f, var elapsedTime: Float = 0f) {
    val remainingTime: Float
        get() = maxTime - elapsedTime

    val isRunning: Boolean
        get() = elapsedTime > 0

    val isElapsed: Boolean
        get() = elapsedTime >= maxTime

    val elapsedPercent: Float
        get() = Math.min(elapsedTime / maxTime, 1f)

    fun reset() {
        elapsedTime = 0f
    }

    fun check(): Boolean {
        val isElapsed = isElapsed
        if (isElapsed) {
            reset()
        }
        return isElapsed
    }

    fun tick(delta: Float) {
        elapsedTime = Math.min(elapsedTime + delta, maxTime)
    }
}
